File: nested Sampling/src/io_utils.py
import os
import csv
import numpy as np

RESULTS_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'results')

METRIC_KEYS = ['logz', 'logzerr', 'logz_resample_std', 'H', 'runtime', 'ncall', 'n_iter', 'eff_nlive', 'nlive_used', 'seed', 'converged']


def save_run(method, dataset, run_id, metrics, phase='test_runs'):
    out_dir = os.path.join(RESULTS_DIR, method, dataset, phase)
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, f'run_{run_id:03d}.csv')
    with open(out_path, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(METRIC_KEYS)
        writer.writerow([metrics.get(k, np.nan) for k in METRIC_KEYS])


def save_summary(method, dataset, all_metrics, phase='test_runs'):
    out_dir = os.path.join(RESULTS_DIR, method, dataset, phase)
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, 'summary.csv')

    keys = METRIC_KEYS
    rows = [['metric', 'mean', 'std', 'min', 'max']]
    for k in keys:
        vals = [m[k] for m in all_metrics if k in m and not np.isnan(m[k])]
        if vals:
            rows.append([k, np.mean(vals), np.std(vals, ddof=1), np.min(vals), np.max(vals)])
        else:
            rows.append([k, np.nan, np.nan, np.nan, np.nan])

    with open(out_path, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerows(rows)

    return out_path


def load_runs(method, dataset, phase='test_runs'):
    out_dir = os.path.join(RESULTS_DIR, method, dataset, phase)
    all_metrics = []
    for fname in sorted(os.listdir(out_dir)):
        if fname.startswith('run_') and fname.endswith('.csv'):
            path = os.path.join(out_dir, fname)
            with open(path, 'r') as f:
                reader = csv.DictReader(f)
                for row in reader:
                    metrics = {k: float(v) for k, v in row.items()}
                    all_metrics.append(metrics)
    return all_metrics

File: Bridge Sampling/src/io_utils.R
METRIC_KEYS <- c("logz", "logzerr", "H", "runtime", "ncall", "n_iter",
                 "eff_nlive", "nlive_used", "seed", "converged")

save_run <- function(dataset, run_id, metrics, phase_dir) {
  out_dir <- file.path(RESULTS_DIR, "bridge_sampling", dataset, phase_dir)
  dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
  out_path <- file.path(out_dir, sprintf("run_%03d.csv", run_id))

  row <- vapply(METRIC_KEYS, function(k) {
    if (is.null(metrics[[k]])) NA_real_ else as.numeric(metrics[[k]])
  }, numeric(1))

  df <- as.data.frame(t(row))
  colnames(df) <- METRIC_KEYS
  write.csv(df, out_path, row.names = FALSE)
  invisible(out_path)
}

save_summary <- function(dataset, all_metrics, phase_dir) {
  out_dir <- file.path(RESULTS_DIR, "bridge_sampling", dataset, phase_dir)
  dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
  out_path <- file.path(out_dir, "summary.csv")

  rows <- data.frame(
    metric = METRIC_KEYS,
    mean   = NA_real_,
    std    = NA_real_,
    min    = NA_real_,
    max    = NA_real_,
    stringsAsFactors = FALSE
  )

  for (k in METRIC_KEYS) {
    vals <- na.omit(vapply(all_metrics, function(m) {
      if (is.null(m[[k]])) NA_real_ else as.numeric(m[[k]])
    }, numeric(1)))
    if (length(vals) > 0) {
      rows[rows$metric == k, c("mean", "std", "min", "max")] <-
        c(mean(vals), if (length(vals) > 1) sd(vals) else 0, min(vals), max(vals))
    }
  }

  write.csv(rows, out_path, row.names = FALSE)
  invisible(out_path)
}

load_bridge_runs <- function(dataset, phase_dir) {
  dir_path <- file.path(RESULTS_DIR, "bridge_sampling", dataset, phase_dir)
  if (!dir.exists(dir_path)) return(list())
  files <- list.files(dir_path, pattern = "^run_\\d+\\.csv$", full.names = TRUE)
  lapply(files, function(f) {
    as.list(read.csv(f))
  })
}
